community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java

```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
     }
 
     @Override
+    public Map<String, Object> getProperties( String... keys )
+    {
+        try ( Statement statement = actions.statement() )
+        {
+            try ( Cursor<NodeItem> node = statement.readOperations().nodeCursor( nodeId ) )
+            {
+                if ( !node.next() )
+                {
+                    throw new NotFoundException( "Node not found",
+                            new EntityNotFoundException( EntityType.NODE, getId() ) );
+                }
+
+                try ( Cursor<PropertyItem> propertyCursor = node.get().properties() )
+                {
+                    Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
+
+                    // Specific properties given
+                    int[] propertyKeys = new int[keys.length];
+                    for ( int i = 0; i < keys.length; i++ )
+                    {
+                        String key = keys[i];
+                        propertyKeys[i] = statement.readOperations().propertyKeyGetForName( key );
+                    }
+
+                    while ( propertyCursor.next() )
+                    {
+                        for ( int i = 0; i < propertyKeys.length; i++ )
+                        {
+                            int propertyKey = propertyKeys[i];
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
+                            }
```

Comment:
Would it be more expensive to have a more dynamic structure for `propertyKeys` and removing the keys that we already found to avoid looping over them again?